gitrepo: Add support for specifying proxy per `GitRepository` (#1109)
fluxcd/source-controller: docs/spec/v1/gitrepositories.md

result in a new Artifact. When the field is set to `false` or removed, it will
resume.

### Proxy

> `.spec.proxy.secretRef.name` is an optional field to specify a name of a Secret
> containing the proxy settings that need to be used for all remote Git operations
> for the particular GitRepository object. The Secret can contain three keys:

Reviewer: ```suggestion
`.spec.proxy.secretRef.name` is an optional field used to specify the name of a Secret
that contains the proxy settings. These settings are required for all remote Git operations
related to the specific GitRepository object.

The Secret can contain three keys:
```

- `address`: The address of the proxy server. This is a required key.
- `username`: The username to use if the proxy server is protected by basic
   authentication. This is an optinal key.
- `password`: The password to use if the proxy server is protected by basic
   authentication. This is an optinal key.

The proxy server must be either HTTP/S or SOCKS5. You can use a SOCKS5 proxy
with a HTTP/S Git repository url.

Examples:

```yaml
---
apiVersion: v1
kind: Secret
metadata:
  name: http-proxy
type: Opaque
stringData:
  address: http://proxy.com
  username: mandalorian
  password: grogu
```

```yaml
---
apiVersion: v1
kind: Secret
metadata:
  name: ssh-proxy
type: Opaque
stringData:
  address: socks5://proxy.com
  username: mandalorian
  password: grogu
```

> Proxying can also be configured in the source-controller pod directly by using

Reviewer: ```suggestion
Proxying can also be configured in the source-controller Pod directly by using
```

result in a new Artifact. When the field is set to `false` or removed, it will
resume.

> ### Proxy

Author: yeah i agree, lets remove it

> ### Proxy

Author: https://github.com/fluxcd/source-controller/pull/1144

result in a new Artifact. When the field is set to `false` or removed, it will
resume.

### Proxy secret reference

`.spec.proxySecretRef.name` is an optional field used to specify the name of a
Secret that contains the proxy settings for the object. These settings are used
for all remote Git operations related to the GitRepository.
The Secret can contain three keys:

- `address`: The address of the proxy server. This is a required key.
- `username`: The username to use if the proxy server is protected by basic
   authentication. This is an optional key.
- `password`: The password to use if the proxy server is protected by basic

>    authentication. This is an optional key.

Author: what difference are you exactly pointing to? is it the period at the prefix? i didn't add them to the proxy docs since they are secret keys and not API fields

resume.

### Proxy secret reference

`.spec.proxySecretRef.name` is an optional field used to specify the name of a
Secret that contains the proxy settings for the object. These settings are used
for all remote Git operations related to the GitRepository.
The Secret can contain three keys:

- `address`, to specify the address of the proxy server. This is a required key.
- `username`, to specify the username to use if the proxy server is protected by
   basic authentication. This is an optional key.
- `password`, to specify the password to use if the proxy server is protected by
   basic authentication. This is an optional key.

The proxy server must be either HTTP/S or SOCKS5. You can use a SOCKS5 proxy
with a HTTP/S Git repository url.

Examples:

```yaml
---
apiVersion: v1
kind: Secret
metadata:
  name: http-proxy
type: Opaque
stringData:
  address: http://proxy.com
  username: mandalorian
  password: grogu
```

```yaml
---
apiVersion: v1
kind: Secret
metadata:
  name: ssh-proxy
type: Opaque
stringData:
  address: socks5://proxy.com
  username: mandalorian
  password: grogu
```

Proxying can also be configured in the source-controller Deployment directly by

> using the standard environment variables such as `HTTPS_PROXY`, `ALL_PROXY`, etc.

Reviewer 2: Maybe add a link to https://fluxcd.io/flux/cheatsheets/bootstrap/#git-repository-access-via-socks5-ssh-proxy?

Reviewer 3: The cheatsheet is going away, we'll have a dedicated page for Proxy settings under `installation/configuration`, see https://github.com/fluxcd/website/issues/1560. For now let's leave it like this, after the docs refactoring we can link to it from here.
